pilot/pkg/model/telemetry.go

```diff
@@ -0,0 +1,146 @@
+	Spec      *tpb.Telemetry `json:"spec"`
+}
+
+// Telemetries organizes Telemetry configuration by namespace.
+type Telemetries struct {
+	// Maps from namespace to the Telemetry configus.
+	NamespaceToTelemetries map[string][]Telemetry `json:"namespace_to_telemetries"`
+
+	// The name of the root namespace.
+	RootNamespace string `json:"root_namespace"`
+}
+
+// GetTelemetries returns the Telemetry configurations for the given environment.
+func GetTelemetries(env *Environment) (*Telemetries, error) {
+	telemetries := &Telemetries{
+		NamespaceToTelemetries: map[string][]Telemetry{},
+		RootNamespace:          env.Mesh().GetRootNamespace(),
+	}
+
+	fromEnv, err := env.List(collections.IstioTelemetryV1Alpha1Telemetries.Resource().GroupVersionKind(), NamespaceAll)
+	if err != nil {
+		return nil, err
+	}
+	sortConfigByCreationTime(fromEnv)
+	for _, config := range fromEnv {
+		telemetry := Telemetry{
+			Name:      config.Name,
+			Namespace: config.Namespace,
+			Spec:      config.Spec.(*tpb.Telemetry),
+		}
+		telemetries.NamespaceToTelemetries[config.Namespace] =
+			append(telemetries.NamespaceToTelemetries[config.Namespace], telemetry)
+	}
+
+	return telemetries, nil
+}
+
+func (t *Telemetries) EffectiveTelemetry(namespace string, workload labels.Collection) *tpb.Telemetry {
+	if t == nil {
+		return nil
+	}
+
+	var effectiveSpec *tpb.Telemetry
+	if t.RootNamespace != "" {
+		effectiveSpec = t.namespaceWideTelemetry(t.RootNamespace)
+	}
+
+	if namespace != t.RootNamespace {
+		nsSpec := t.namespaceWideTelemetry(namespace)
+		if nsSpec != nil && effectiveSpec != nil {
+			effectiveSpec = shallowMerge(effectiveSpec, nsSpec)
+		}
+	}
+
+	for _, telemetry := range t.NamespaceToTelemetries[namespace] {
+		spec := telemetry.Spec
+		if len(spec.GetSelector().GetMatchLabels()) == 0 {
+			continue
+		}
+		selector := labels.Instance(spec.GetSelector().GetMatchLabels())
+		if workload.IsSupersetOf(selector) {
+			effectiveSpec = shallowMerge(effectiveSpec, spec)
+		}
+	}
+
+	return effectiveSpec
+}
+
+func (t *Telemetries) namespaceWideTelemetry(namespace string) *tpb.Telemetry {
+	for _, tel := range t.NamespaceToTelemetries[namespace] {
+		spec := tel.Spec
+		if len(spec.GetSelector().GetMatchLabels()) == 0 {
+			return spec
+		}
+	}
+	return nil
+}
+
+func shallowMerge(parent, child *tpb.Telemetry) *tpb.Telemetry {
+	return shallowMergeTracing(parent, child)
+}
+
+func shallowMergeTracing(parent, child *tpb.Telemetry) *tpb.Telemetry {
+	merged := parent.DeepCopy()
+
+	if len(child.GetTracing()) == 0 {
+		return merged
+	}
+
+	childCopy := child.DeepCopy()
+
+	if len(merged.GetTracing()) == 0 {
+		merged.Tracing = childCopy.Tracing
+		return merged
+	}
+
+	// only use the first Tracing for now (all that is suppported)
+	childTracing := childCopy.Tracing[0]
+	mergedTracing := merged.Tracing[0]
+	if len(childTracing.Providers) != 0 {
+		mergedTracing.Providers = childTracing.Providers
+	}
+
+	if len(childTracing.CustomTags) != 0 {
```

Comment:
Do we need to check if customTags is nil or not? If it is nil, that is child expressing no custom tags right?

Reply:
Good point re: overriding to turn off. This doesn't provide a way to opt out of child tags from your parent unless you are specifying your own replacement overrides. I'll take another look. We may want something that involves API changes for consistency with other fields.

Reply:
So if child span does not set customTags, it will result in an empty slice here instead of nil?